cpp.mk

```diff
@@ -79,11 +79,24 @@ CLEAN_DIRS += $${$1_OBJECT_DIR}
 clangtidy_$1: $${$1_CLANGTIDY}
 	cat $${$1_CLANGTIDY}
 
+# Service tests must run under Docker, to provide isolation, and this Makefile
+# enforces this. To run service tests, repositories must provide:
+# - a service_tests/ directory at the top level
+# - a Dockerfile in that directory which will run the tests
+# - a service_test_copy_files_NAME Makefile target which is responsible for
+#   copying any necessary files (e.g. the binary and libraries under test) into
+#   the service_test/ directory structure - the Dockerfile cannot see anything
+#   above this directory.
+
+.PHONY: service_test_copy_files_$1
+service_test_copy_files_$1:
```

Comment:
I don't think you actually need this target - users could just chain from `service_test_$1` directly. Happy either way though.